java-checks/src/main/java/org/sonar/java/checks/helpers/Javadoc.java

```diff
@@ -0,0 +1,266 @@
+        .map(Javadoc::exceptionName)
+        .filter(Objects::nonNull)
+        .collect(Collectors.toList());
+    } else if (tree.is(CLASS_KINDS)) {
+      elementParameters = ((ClassTree) tree).typeParameters().stream()
+        .map(TypeParameterTree::identifier)
+        .map(IdentifierTree::name)
+        .map(name -> "<" + name + ">")
+        .collect(Collectors.toList());
+      elementExceptionNames = Collections.emptyList();
+    } else {
+      elementParameters = Collections.emptyList();
+      elementExceptionNames = Collections.emptyList();
+    }
+
+    List<String> javadocLines = cleanLines(PublicApiChecker.getApiJavadoc(tree));
+    mainDescription = getDescription(javadocLines, -1, "");
+    blockTagDescriptions = extractBlockTags(javadocLines, Arrays.asList(BlockTag.values()));
+  }
+
+  public boolean noMainDescription() {
+    return isEmptyDescription(mainDescription);
+  }
+
+  public boolean noReturnDescription() {
+    return isEmptyDescription(blockTagDescriptions.get(BlockTagKey.of(BlockTag.RETURN, null)));
+  }
+
+  public Set<String> undocumentedParameters() {
+    return elementParameters.stream()
+      .filter(name -> isEmptyDescription(blockTagDescriptions.get(BlockTagKey.of(BlockTag.PARAM, name))))
+      .collect(Collectors.toSet());
+  }
+
+  public Set<String> undocumentedThrownExceptions() {
+    Map<String, List<String>> thrownExceptionsMap = getJavadocExceptions();
+    List<String> exceptionNames = elementExceptionNames;
+    if (exceptionNames.size() == 1 && GENERIC_EXCEPTIONS.contains(toSimpleName(exceptionNames.get(0))) && !thrownExceptionsMap.isEmpty()) {
+      // check for documented exceptions without description when only "Exception" is declared as being thrown
+      return thrownExceptionsMap.entrySet().stream()
+        .filter(e -> isEmptyDescription(e.getValue()))
+        .map(Map.Entry::getKey)
+        .map(Javadoc::toSimpleName)
+        .collect(Collectors.toSet());
+    }
+    return exceptionNames.stream()
+      .map(Javadoc::toSimpleName)
+      .filter(simpleName -> noDescriptionForException(thrownExceptionsMap, simpleName))
+      .collect(Collectors.toSet());
+  }
+
+  private boolean noDescriptionForException(Map<String, List<String>> thrownExceptionsMap, String exceptionSimpleName) {
+    List<String> descriptions = thrownExceptionsMap.get(exceptionSimpleName);
+    if (descriptions == null) {
+      // exceptions used in javadoc is using fully qualified name when method declaration use simple name
+      descriptions = thrownExceptionsMap.entrySet().stream()
+        .filter(e -> toSimpleName(e.getKey()).equals(exceptionSimpleName))
+        .map(Map.Entry::getValue)
+        .flatMap(List::stream)
+        .collect(Collectors.toList());
+    }
+    return isEmptyDescription(descriptions);
+  }
+
+  private Map<String, List<String>> getJavadocExceptions() {
+    return blockTagDescriptions.entrySet().stream()
```

Comment:
probably worth being computed directly when creating the `JavaDoc` object, or at least only once if it needs to be called multiple times.